CASMPET-7448 add documentation for booting storage node into new image (#5846)
Cray-HPE/docs-csm: operations/node_management/Boot_storage_node_into_new_image.md

> Note: In CSM 1.7, this procedure is no longer needed. Instead, a normal storage node upgrade will pull the correct image even if `rd.live.dir` did not change.

## Procedure overview

This procedure will upgrade storage nodes using the normal Argo workflow framework.
Prior to upgrading the storage node, the `rd.live.dir` will be set to the `STORAGE_NODE_IMAGE_ID`. Because this is a unique value, the node image will be pulled successfully during `cloud-init`.

This procedure can be followed prior to booting a storage node into a new image or
if a storage node has already attempted to boot into a new image and it failed in `cloud-init` due to `rd.live.dir` already existing.

If the storage node is currently being upgraded and `cloud-init` failed, it is best to leave the Argo Workflow running while following the procedure below.
It is okay if the workflow has already been stopped, a step will provide instructions for resuming it.

## Procedure

> 1. (`ncn-m001#`) Set `TARGET_NCN` to the storage node hostname for the node that will be booted into a new image

Reviewer: ```suggestion
1. (`ncn-m001#`) Set `TARGET_NCN` to the storage node hostname for the node that will be booted into a new image.
```

    ```bash
    TARGET_NCN=ncn-s00X
    ```

> 1. (`ncn-m001#`) Set the image id of the image that the node should be booted into

Reviewer: ```suggestion
1. (`ncn-m001#`) Set the image ID of the image that the node should be booted into.
```

    ```bash
    STORAGE_NODE_IMAGE_ID=<image_id>
    ```

> 1. (`ncn-m001#`) Set `TOKEN` and `TARGET_XNAME` variables. These variables are needed in subsequent steps.

Reviewer: ```suggestion
1. (`ncn-m001#`) Set the `TOKEN` and `TARGET_XNAME` variables. These variables are needed in subsequent steps.
```

    ```bash
    export TOKEN=$(curl -k -s -S -d grant_type=client_credentials \
    -d client_id=admin-client \
    -d client_secret=$(kubectl get secrets admin-client-auth -o jsonpath='{.data.client-secret}' | base64 -d) \
    https://api-gw-service-nmn.local/keycloak/realms/shasta/protocol/openid-connect/token | jq -r '.access_token')

    TARGET_XNAME=$(curl -s -k -H "Authorization: Bearer ${TOKEN}" "https://api-gw-service-nmn.local/apis/sls/v1/search/hardware?extra_properties.Role=Management" \
    | jq -r ".[] | select(.ExtraProperties.Aliases[] | contains(\"$TARGET_NCN\")) | .Xname")
    echo "TARGET_XNAME: $TARGET_XNAME"
    ```

1. (`ncn-m001#`) Set the `rd.live.dir` on the `TARGET_NCN` to the `STORAGE_NODE_IMAGE_ID`. This ensures `rd.live.dir` will be a unique value.

    ```bash
    csi handoff bss-update-param --delete rd.live.dir --limit "${TARGET_XNAME}"
    csi handoff bss-update-param --set rd.live.dir=${STORAGE_NODE_IMAGE_ID} --limit "${TARGET_XNAME}"
    ```

> 1. (`ncn-m001#`) Follow one of the two options below
> 
>     1. **Option 1:** Start a new storage node upgrade if one has not already been started
> 
>         > Note: if a new CFS configuration is also necessary, append the argument `--desired-cfs-conf <configuration>` to the command below.
> 
>         ```bash
>         /usr/share/doc/csm/upgrade/scripts/upgrade/ncn-upgrade-worker-storage-nodes.sh $TARGET_NCN --upgrade --image-id $STORAGE_NODE_IMAGE_ID
>         ```
> 
>     1. **Option 2:** If a storage node has already attempted to boot into a new image and `cloud-init` has failed, manually power cycle the node.

Reviewer: ```suggestion
1. (`ncn-m001#`) Follow the steps for one of the two options depending on the status of the storage node.

    * **Option 1:** Start a new storage node upgrade if one has not already been started.

        > Note: if a new CFS configuration is also necessary, append the argument `--desired-cfs-conf <configuration>` to the command below.

        ```bash
        /usr/share/doc/csm/upgrade/scripts/upgrade/ncn-upgrade-worker-storage-nodes.sh $TARGET_NCN --upgrade --image-id $STORAGE_NODE_IMAGE_ID
        ```

    * **Option 2:** If a storage node has already attempted to boot into a new image and `cloud-init` has failed, manually power cycle the node.
```

# Boot a storage node into new image without upgrading CSM

## Purpose

This procedure is needed anytime a storage node is booted into a new image except for during a CSM upgrade.

> For example, if a storage node image been rebuilt, then this procedure is necessary to boot storage nodes into the newly built image.

Reviewer 2: "if a storage node image *has* been rebuilt" maybe?

This procedure is not necessary when upgrading a storage node from one CSM major release to the next major release.

> Instead of following this procedure to upgrade the storage nodes into the new image, the storage nodes could also be rebuilt into the new image.

Reviewer 2: I think this is kind of confusing.  In the "For example" statement above it is stated that this procedure is to boot storage nodes into the newly built image.  Now it's saying that this procedure is to upgrade the storage node into a new image and that there is another procedure for the storage nodes to be "rebuilt into the new image". 

I'm confused.  Is this procedure "booting" a storage node into a new image, "upgrading" the storage nodes into a new image, or "rebuilding" the storage nodes into a new image.  Title says "Boot".

Author: I have tried to clarify this.
- The storage node is booted into a new image
- It is boot into a new image by using the storage node upgrade Argo workflow
- The storage node could be rebuilt which would wipe the node but that is not the process described here

Because the storage node is not wiped, the `rd.live.dir` still exists when the node is rebooted.
This is due to a change in how `rd.live.overlay.reset` works in CSM 1.5 and 1.6.
If the `rd.live.dir` is not changed when booting a storage node to a new node image, the new image will not be pulled since the `rd.live.dir` still exists.
This can cause `cloud-init` to fail on the storage node or the node will boot but will use the previous `/run/initramfs/live/<CSM_version>` directory that still contains the previous image.
To avoid this issue and ensure the new node image is pulled, the `rd.live.dir` should be set to unique value.
In this procedure, the `STORAGE_NODE_IMAGE_ID` is used as the unique value.

This issue is not hit when upgrading from one CSM major release to the next CSM major release because the `rd.live.dir` is changed on all NCNs to the new CSM major release version. This allows the node image to be pulled successfully during `cloud-init`.

> Note: In CSM 1.7, this procedure is no longer needed. Instead, a normal storage node upgrade will pull the correct image even if `rd.live.dir` did not change.

## Procedure overview

> This procedure will upgrade storage nodes using the normal Argo workflow framework.

Reviewer 2: upgrade = boot with a new image here?
